feat: add cinnamon

mkg20001 opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

I need this for myself and I think a lot of people want this, too

**NOTE: THIS PR IS WIP**
:exclamation:  It's been decided that most expressions will be moved out of this PR, and into individual ones.

## The packages here are being continued to be developed in https://github.com/mkg20001/nixpkgs and will be moved to PRs to get them ready for upstream

#### Also, I'm running this on my laptop already. So it works and you can try it already, it's just a bit broken at this point...


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

mkg20001 committed:
feat: add cinnamon

mkg20001 committed:
feat: enable cinnamon

mkg20001 committed:
feat: add new and old cinnamon port

mkg20001 commented:
I've added the old cinnamon port from the repo before removal & the new cinnamon port https://github.com/cawilliamson/nixos-cinnamon/ from @cawilliamson 

Going to merge them together and make a new port, latest going to remove the leftover files

mkg20001 committed:
feat: cinnamon scope

mkg20001 committed:
refactor: move cinnamon new into cinnamon target

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/porting-cinnamon-to-nixos/3512/7

worldofpeace commented:
I'd be happy to review this :+1:

mkg20001 commented:
I've added `services.xserver.desktopManager.cinnamon.enable` as an option and referenced cinnamon.nix in `desktop-manager/default.nix`, yet my local nixbuild fails (evaluation logs show it seems to use my cloned nixpkgs)

mkg20001 committed:
feat: add myself as maintainer

mkg20001 committed:
feat: make it build

mkg20001 committed:
Revert "spidermonkey_52: drop, as is unused"

This reverts commit 18d7929e95a76210d771c1648b97d90737756854.

mkg20001 committed:
feat: hooks and fixes

mkg20001 committed:
feat: fixesnstuff

mkg20001 committed:
fix: more patches

mkg20001 committed:
fix: hash

mkg20001 committed:
fix: hash for real

worldofpeace commented:
> I've added `services.xserver.desktopManager.cinnamon.enable` as an option and referenced cinnamon.nix in `desktop-manager/default.nix`, yet my local nixbuild fails (evaluation logs show it seems to use my cloned nixpkgs). Any idea?

I think the module might be a far off idea just yet. I'd probably try to get the packaging very good first and maybe just use some configuration locally to verify their installation. So, I could much better assist with packaging first, as the module might prove more difficult. (but I can for sure help with this as well).

worldofpeace commented on pkgs/desktops/cinnamon/cjs/default.nix:
```diff
@@ -0,0 +1,33 @@
+    sha256 = "1yvzvfm1z0d8ca9vk9s0gbsir0ln7mcmlbczf0hh8vzpsg7m1zk5";
+  };
+
+  buildInputs = [ autoconf-archive dbus_glib gobjectIntrospection pkgconfig spidermonkey_52 ];
```
It seems it may be possible for this to build with a later spidermonkey version
* https://github.com/linuxmint/cjs/blob/fe3b1ec434700f45bf75615285a18decfa3aca1a/NEWS#L360

jtojnar commented on pkgs/desktops/cinnamon-old/cinnamon-control-center.nix:
```diff
@@ -0,0 +1,41 @@
+    sha256 = "0kivqdgsf8w257j2ja6fap0dpvljcnb9gphr3knp7y6ma2d1gfv3";
+  };
+
+ configureFlags = "--enable-systemd --disable-update-mimedb" ;
```
`configureFlags` need to be a list.

jtojnar commented on pkgs/desktops/cinnamon-old/cinnamon-control-center.nix:
```diff
@@ -0,0 +1,41 @@
+  version = "2.0.9";
+in
+stdenv.mkDerivation {
+  name = "cinnamon-control-center-${version}";
```
```suggestion
  pname = "cinnamon-control-center";
  inherit version;
```

jtojnar commented on pkgs/desktops/cinnamon-old/cinnamon-control-center.nix:
```diff
@@ -0,0 +1,41 @@
+stdenv.mkDerivation {
+  name = "cinnamon-control-center-${version}";
+
+  src = fetchurl {
```
Use `fetchFromGitHub` when building from the archive generated by GitHub.

jtojnar commented on pkgs/desktops/cinnamon-old/cinnamon-control-center.nix:
```diff
@@ -0,0 +1,41 @@
+  patches = [ ./region.patch];
+
+  buildInputs = [
+    pkgconfig autoreconfHook
```
Programs executed during build (`pkgconfig autoreconfHook  gettext gnome_common intltool libxslt`) should go to `nativeBuildInputs`.

jtojnar commented on pkgs/desktops/cinnamon-old/cinnamon-control-center.nix:
```diff
@@ -0,0 +1,41 @@
+
+  meta = {
+    homepage = "http://cinnamon.linuxmint.com";
+    description = "The cinnamon session files" ;
```
This description does not seem to match.

jtojnar commented on pkgs/top-level/aliases.nix:
```diff
@@ -365,7 +365,6 @@ mapAliases ({
   speedtest_cli = speedtest-cli;  # added 2015-02-17
   spice_gtk = spice-gtk; # added 2018-02-25
   spice_protocol = spice-protocol; # added 2018-02-25
-  spidermonkey_52 = throw "spidermonkey_52 has been removed. Please use spidermonkey_60 instead."; # added 2019-10-16
```
Firefox 52 has not been receiving security updates for over a year. Because this is security critical component, its re-introduction is not a good idea. Could you ask upstream to switch to a newer Spidermonkey version?

worldofpeace replied:
https://github.com/NixOS/nixpkgs/pull/74242#discussion_r350716404

mkg20001 replied:
55 isn't an ESR, afaik? Wouldn't that be worse, then?

worldofpeace replied:
I've done some research and it doesn't seem any distro is currently using a newer spidermonkey version. I think we should ask upstream.

mkg20001 replied:
https://github.com/linuxmint/cjs/issues/80

mkg20001 replied:
Upstream https://github.com/linuxmint/cjs/issues/80

Should we continue with adding spidermonkey 52 if it doesn't get resolved in time?

mkg20001 replied:
Upstream said 4.4.0: wontfix, 4.6.0: likely, 4.8.0: definitively.
Since 4.4.0 came out basically this week, we can assume that we'll either have to wait until 4.6.0 or add 52 for 4.4.0 one last time.

worldofpeace replied:
Sounds good to me.

jtojnar commented on pkgs/desktops/nixos-cinnamon/default.nix:
```diff
@@ -0,0 +1,12 @@
+  cinnamon = super.callPackage ./cinnamon { };
+  cinnamon-desktop = super.callPackage ./cinnamon-desktop { };
+  cinnamon-menus = super.callPackage ./cinnamon-menus { };
+  cinnamon-session = super.callPackage ./cinnamon-session { gconf = super.gnome2.GConf; };
```
Does this still require GConf? It has been deprecated over 8 years ago.

mkg20001 replied:
Apparently. I've had to add it again.

jtojnar commented on pkgs/desktops/cinnamon/xapps/default.nix:
```diff
@@ -0,0 +1,41 @@
+  };
+
+  patches = [
+    ./py-override.patch
```
Could you add comment with a source and explanation?

worldofpeace replied:
I believe it's a patch I wrote for @cawilliamson from https://github.com/linuxmint/xapps/pull/65.

mkg20001 replied:
@worldofpeace What does the patch do (what should I add as comment)?

worldofpeace replied:
I'd just add a comment with the link to the PR. It's all described there.

mkg20001 commented:
> I think the module might be a far off idea just yet. I'd probably try to get the packaging very good first and maybe just use some configuration locally to verify their installation. So, I could much better assist with packaging first, as the module might prove more difficult. (but I can for sure help with this as well).

Can we keep it now in this PR and I'll move it to another PR later, so we can merge just the packages in this PR?

worldofpeace commented:
> > I think the module might be a far off idea just yet. I'd probably try to get the packaging very good first and maybe just use some configuration locally to verify their installation. So, I could much better assist with packaging first, as the module might prove more difficult. (but I can for sure help with this as well).
> 
> Can we keep it now in this PR and I'll move it to another PR later, so we can merge just the packages in this PR?

I believe you're suggesting merging packaging first and then a nixos module?
That sounds right, it's an approach we've done for packaging deepin. It would make the reviews smaller and easier to manage.

mkg20001 committed:
feat: more package fixes

mkg20001 committed:
fix: cinnamon translations

mkg20001 committed:
fix: xapps can't find g-codegen

mkg20001 committed:
misc: cleanup old file

mkg20001 committed:
feat: rewrite cinnamon-control-center package

mkg20001 committed:
fix: more fixes

mkg20001 committed:
fix: cinnamon

mkg20001 commented:
I have trouble adding libnm as a dependency for cinnamon-control-center

I can't find any headers in that package. Are they exposed somewhere else than `networkmanager`?

mkg20001 committed:
fix: use right path to cinnamon session

mkg20001 committed:
feat: upgrade cjs

mkg20001 committed:
fix: control-center build

mkg20001 committed:
feat: add cinnamon session & desktop as sys packages

mkg20001 committed:
fix: finally fix cinnamon control center

mkg20001 committed:
fix: cinnamon build

mkg20001 committed:
fix: link share as well to fix cinnamon session

mkg20001 committed:
fix: tzdata hardcoded paths

mkg20001 committed:
fix: datetime/backward missing

worldofpeace commented on pkgs/desktops/cinnamon/cinnamon-menus/default.nix:
```diff
@@ -0,0 +1,24 @@
+  };
+
+  buildInputs = [ glib gobjectIntrospection pkgconfig ];
+  nativeBuildInputs = [ meson ninja wrapGAppsHook cmake ];
```
```suggestion
  nativeBuildInputs = [ meson ninja wrapGAppsHook ];
```

worldofpeace replied:
You should also set the following meson flags
```
mesonFlags = [
  "enable_debug=false"
  "deprecated_warnings=false"
];
```

The deprecated warnings are going to probably cause issues in nixpkgs.

worldofpeace commented on pkgs/desktops/cinnamon/cinnamon-menus/default.nix:
```diff
@@ -0,0 +1,24 @@
+    sha256 = "0q4qj28swi2y93fj7pfil68l2cf9gmhbk6jmr8d70l54xf7sigsh";
+  };
+
+  buildInputs = [ glib gobjectIntrospection pkgconfig ];
```
`gobjectIntrospection` should be in `nativeBuildInputs`.
Also, this is an old alias we don't use anymore, please use the `gobject-introspection` attribute.

worldofpeace commented on pkgs/desktops/cinnamon/cinnamon-menus/default.nix:
```diff
@@ -0,0 +1,24 @@
+
+  src = fetchFromGitHub {
+    owner = "linuxmint";
+    repo = "${pname}";
```
```suggestion
    repo = pname;
```

worldofpeace commented on pkgs/desktops/cinnamon/cinnamon-menus/default.nix:
```diff
@@ -0,0 +1,24 @@
+  src = fetchFromGitHub {
+    owner = "linuxmint";
+    repo = "${pname}";
+    rev = "${version}";
```
```suggestion
    rev = version;
```

worldofpeace commented on pkgs/desktops/cinnamon/cinnamon-menus/default.nix:
```diff
@@ -0,0 +1,24 @@
+    description = "The cinnamon-menu library";
+
+    platforms = stdenv.lib.platforms.linux;
+    maintainers = [ stdenv.lib.maintainers.mkg20001 ];
```
A preferred style in nixpkgs is to do `meta with stdenv.lib; {` and don't use the `stdenv.lib` attr path here.

worldofpeace commented on pkgs/desktops/cinnamon/cinnamon-menus/default.nix:
```diff
@@ -0,0 +1,24 @@
+  nativeBuildInputs = [ meson ninja wrapGAppsHook cmake ];
+
+  meta = {
+    homepage = "http://cinnamon.linuxmint.com";
```
A webpage that's still valid that's similar to this is https://developer.linuxmint.com/projects/cinnamon-projects.html.
Or maybe just use the url to github.

worldofpeace commented on pkgs/desktops/cinnamon/cinnamon-menus/default.nix:
```diff
@@ -0,0 +1,24 @@
+
+  meta = {
+    homepage = "http://cinnamon.linuxmint.com";
+    description = "The cinnamon-menu library";
```
This description isn't great
```suggestion
    description = "A menu system for the Cinnamon project";
```

it's also nixpkgs convention to not include the software's name in the description.

worldofpeace reviewed (changes requested):
I've decided to review this a package at a time, so the following comments are just for `cinnamon-menus`.

mkg20001 committed:
fix: control-center tests

mkg20001 committed:
fix: menu lib desc

Co-Authored-By: worldofpeace <worldofpeace@protonmail.ch>

mkg20001 committed:
fix: nixos style

mkg20001 committed:
fix: metadata

mkg20001 committed:
fix: add cinnamon to cinnamon de

mkg20001 committed:
fix: rename cinnamon to cinnamon-common

mkg20001 committed:
feat: add cinnamon-screensaver

mkg20001 committed:
feat: add nemo

mkg20001 committed:
feat: add nm-applet

worldofpeace reviewed (changes requested):
I'm seeing you're doing a lot of 
```
preConfigurePhase = ...

confFixPhase = ..
```
and what you're actually doing here is patching the sources.

What you should instead should be doing, is is done frequently in many other expressions, is to use `postPatch`

example:
```
postPatch = ''
  patchShebangs autogen.sh
'';
```

Also note, you don't need the `./` because you're not executing anything.

worldofpeace commented on pkgs/desktops/cinnamon/cinnamon-session/default.nix:
```diff
@@ -0,0 +1,51 @@
+    patchShebangs data/meson_install_schemas.py
+
+    substituteInPlace cinnamon-session/meson.build \
+      --replace "# elogind" "dbus_glib"
```
Apparently they're completely ported to gdbus https://github.com/linuxmint/cinnamon-session/commit/16371b437e7bb98610bca2c814d43491a5d65173, but I think there's one they missed https://github.com/linuxmint/cinnamon-session/blob/master/cinnamon-session/csm-util.c#L35.

worldofpeace replied:
Apparently they're completely ported to gdbus https://github.com/linuxmint/cinnamon-session/commit/16371b437e7bb98610bca2c814d43491a5d65173, but I think there's one they missed https://github.com/linuxmint/cinnamon-session/blob/master/cinnamon-session/csm-util.c#L35.

worldofpeace commented on pkgs/desktops/cinnamon/cinnamon-session/default.nix:
```diff
@@ -0,0 +1,51 @@
+        --suffix XDG_DATA_DIRS : "${cinnamon-desktop}/share" \
+        --suffix XDG_DATA_DIRS : "${cinnamon-settings-daemon}/etc/xdg"
+    done
+  '';
```
This shouldn't be needed.

However you do need to add `cinnanmon-settings-daemon` for xdg autostarts.
We do the same for gnome-session
* https://github.com/NixOS/nixpkgs/blob/d7b63dc4cef96b184aa29bd136704b8b22c9b838/pkgs/desktops/gnome-3/core/gnome-session/default.nix#L50

see https://nixos.org/nixpkgs/manual/#ssec-gnome-hooks ` You can also pass additional arguments to makeWrapper using gappsWrapperArgs in preFixup hook:`

mkg20001 replied:
What should I change here then? `postFixUp` => `preFixUp` ?

worldofpeace commented on pkgs/desktops/cinnamon/cinnamon-session/default.nix:
```diff
@@ -0,0 +1,51 @@
+  Run-time dependency libsystemd-login found: NO (tried pkgconfig and cmake)
+  Run-time dependency libsystemd found: NO (tried pkgconfig and cmake) */
+
+  buildInputs = [ cinnamon-desktop cinnamon-settings-daemon dbus-glib docbook_xsl docbook_xml_dtd_412 glib gsettings_desktop_schemas gtk3 libcanberra libxslt makeWrapper pkgconfig python3 xapps xmlto xorg.xtrans xorg.libXtst gnome2.GConf systemd ];
```
it's possible  docbook_xsl docbook_xml_dtd_412 python3 should be native.

worldofpeace replied:
gsettings_desktop_schemas is an alias to gsettings-desktops-schemas

jtojnar replied:
Maybe add `nixpkgs.config.allowAliases = false;` to your `configuration.nix`, or `allowAliases = false;` to your `~/.config/nixpkgs/config.nix`.

mkg20001 commented:
Nemo seems to be producing a derivation with all files being stored in `$out/$out` (/nix/store/hash-nemo/nix/store/hash-nemo)

Clueless as to why that happens. Any ideas?

Also still can't get cinnamon-control-center to find network manager during configure.

worldofpeace reviewed (changes requested):
This is for cinnamon-session.

It would also be great if you could incrementally mimic a style like
* https://github.com/NixOS/nixpkgs/blob/master/pkgs/desktops/gnome-3/core/gnome-settings-daemon/default.nix

where the list items and function formal arguments are one per line.

mkg20001 committed:
fix: confFixPhase => postPatch & gobjectIntrospection alias fix

jtojnar commented on pkgs/desktops/cinnamon/nemo/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  preInstallPhases = "exportDestdir";
+
+  exportDestdir = "export DESTDIR=$out";
```
DESTDIR is not a good idea https://github.com/NixOS/nixpkgs/issues/65718

mkg20001 replied:
It's used in an if statement to determine whether or not update-mimedb should be run by some script. Should I add a comment or add a patch?

jtojnar replied:
Comment on anything out of ordinary is a good idea.

mkg20001 replied:
Looks like it breaks stuff anyways, made a patch to remove script instead

jtojnar commented on pkgs/desktops/cinnamon/nemo/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  exportDestdir = "export DESTDIR=$out";
+
+  NIX_CFLAGS_COMPILE = [ "-I${glib.dev}/include/gio-unix-2.0" ];
```
Could you add a comment with link to upstream issue as per https://github.com/NixOS/nixpkgs/issues/36468

mkg20001 replied:
Added and commented on that issue

worldofpeace commented:
> Nemo seems to be producing a derivation with all files being stored in `$out/$out` (/nix/store/hash-nemo/nix/store/hash-nemo)
> 
> Clueless as to why that happens. Any ideas?
> 
> Also still can't get cinnamon-control-center to find network manager during configure.


This is likely because of https://github.com/NixOS/nixpkgs/pull/74242#discussion_r351931504.

jtojnar commented on pkgs/desktops/cinnamon/nemo/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  patches = [ ./0001-patch-datadir.patch ];
+
+  preInstallPhases = "exportDestdir";
```
This should be a list (preparation for `_structuredAttrs`). Or just use `preInstall`.

mkg20001 committed:
fix: nemo dbl $out & DESTDIR usage

mkg20001 commented:
>> Nemo seems to be producing a derivation with all files being stored in $out/$out (/nix/store/hash-nemo/nix/store/hash-nemo)

>> Clueless as to why that happens. Any ideas?

>> Also still can't get cinnamon-control-center to find network manager during configure.

> This is likely because of #74242 (comment).

Ok, fixed. The cinnamon-control-center issue seems to not be affected by DESTDIR

worldofpeace commented on pkgs/desktops/cinnamon/nemo/0002-patch-remove-update-mime-db-script.patch:
```diff
@@ -0,0 +1,27 @@
+-
+-if not os.environ.get('DESTDIR'):
+-    print('Updating mime database...')
+-    subprocess.call(['update-mime-database', mimedir])
```
You just need to add `shared-mime-info` to `nativeBuildInputs`. Why does this not work? I recall packaging things with this exact script running this.

mkg20001 replied:
In the old nixos cinnamon there was a package that did "--disable-update-mimedb" via configureFlags, so I assumed this update step isn't necessary, is it?

worldofpeace replied:
It isn't but you now have to maintain this patch. That dependency is already in the runtime closure so you don't lose much by having it ran.

mkg20001 commented:
cinnamon-screensaver complains it can't find the python module gi. What's the nixos package name for that?

mkg20001 committed:
chore: upgrade xapps

mkg20001 committed:
refactor: deps in cinnamon-session

mkg20001 committed:
fix: add review changes

jtojnar commented:
> cinnamon-screensaver complains it can't find the python module gi. What's the nixos package name for that?

`python3.pkgs.pygobject3`

mkg20001 committed:
chore: upgrade deps

mkg20001 commented:
> python3.pkgs.pygobject3

Still `ModuleNotFoundError: No module named 'gi'`

mkg20001 committed:
fix: use non-aliases

mkg20001 committed:
fix: gobject introspection

jtojnar commented on pkgs/desktops/cinnamon/cinnamon-screensaver/default.nix:
```diff
@@ -0,0 +1,37 @@
+{ stdenv, fetchFromGitHub, pkgconfig, autoreconfHook, glib, gettext, cinnamon-desktop, intltool, libxslt, gtk3, libnotify,
+gnome-menus, libxml2, systemd, upower, cinnamon-settings-daemon, colord, polkit, ibus, libcanberra_gtk3, libpulseaudio, isocodes, kerberos,
```
```suggestion
```

jtojnar commented:
> > python3.pkgs.pygobject3
> 
> Still `ModuleNotFoundError: No module named 'gi'`

The easiest way to ensure the package is included is to use python with the package: `(python3.withPackages (pp: with pp; [ pygobject3 ]))`

mkg20001 committed:
refactor: cleanup pkgs/desktops/cinnamon/cinnamon-screensaver/default.nix

Co-Authored-By: Jan Tojnar <jtojnar@gmail.com>

mkg20001 committed:
fix: use py3 with pkgs

mkg20001 committed:
fix: add setproctitle as dep to cinnamon-screensaver

jtojnar commented:
Also there seems to be hardcoded `/usr/share/locale` in `/nix/store/226b0mf5srrircrrzg27vm1a19l4inkz-cinnamon-screensaver-4.4.0/share/cinnamon-screensaver/cinnamon-screensaver-main.py`

mkg20001 committed:
fix: remove another alias

mkg20001 commented:
cinnamon-screensaver now needs xapp (from xapps) packaged as python package

mkg20001 committed:
fix: remove more aliases

mkg20001 committed:
fix: apply review comments

mkg20001 committed:
fix: hardcoded path

mkg20001 committed:
feat: package xapp python

mkg20001 committed:
fix: cinnamon killer bad interpreter

mkg20001 committed:
fix: add dep

mkg20001 committed:
fix: cinnamon-screensaver deps

mkg20001 commented:
Crashes
- [ ] cinnamon-screensaver: ValueError: Namespace AccountsService not available
- [ ] cinnamon-common cinnamon-launcher-wrapped: FileNotFoundError (main.js)
- [ ] cinnamon-common: JS ERROR: ImportError: No JS module 'cairoNative' found in search path
- [ ] cinnamon-common cinnamon-killer-daemon-wrapped: ValueError: Namespace Keybinder not available

mkg20001 committed:
feat: add adwaita icon theme

mkg20001 committed:
chore: add missing deps

mkg20001 committed:
fix: add xapps dep

mkg20001 committed:
fix: dep

mkg20001 committed:
fix: add killall as dep

mkg20001 committed:
feat: use cinnamon-session-cinnamon instead of cinnamon-session

mkg20001 committed:
feat: add 2d session

mkg20001 committed:
fix: add pycairo

mkg20001 committed:
fix: add cairo

mkg20001 committed:
fix: cairo bug

mkg20001 committed:
fix: typo

mkg20001 committed:
feat: add caribou

mkg20001 committed:
typo

mkg20001 committed:
fix: cinnamon

mkg20001 commented:
![screenshot](https://user-images.githubusercontent.com/7735145/69907656-c0fea900-13d8-11ea-91a1-8d945dd17f38.png)

It works!
The solution was to add caribou to both cjs and cinnamon-common

mkg20001 commented:
![screenshot](https://user-images.githubusercontent.com/7735145/69907747-2d7aa780-13db-11ea-9f33-4c5a37d454c0.png)

Terminal looks messed up, firefox works

jtojnar commented:
I had that when no fonts are installed. Try

```nix
fonts.enableDefaultFonts = true
```

mkg20001 committed:
fix: cinnamon fonts

mkg20001 commented:
That didn't quite work, instead I just copied some code from gnome3 and added it to the module
Now it works

mkg20001 commented:
In which package is the adwaita theme? Can't find it anywhere, cinnamon depends on it as it's the default theme.

jtojnar commented:
Adwaita GTK theme is built into GTK.

mkg20001 committed:
fix: theme & icons

mkg20001 commented:
This PR should be ready now.

mkg20001 commented:
![screenshot](https://user-images.githubusercontent.com/7735145/69909581-2e243580-13fd-11ea-8d8b-5a4f7d39c0c9.png)

Cinnamon, with fonts, theme & icons fixed

worldofpeace commented:
Ok, there's still quite a few issues here. Mostly when it comes to the gobject-introspection using packages. I should have some time this week or later to review just those.

While I'm here, why's there a `cinnamon-old` directory?

worldofpeace commented on nixos/modules/services/x11/desktop-managers/cinnamon.nix:
```diff
@@ -0,0 +1,79 @@
+        '';
+      }
+    ];
+  };
```
I'm assuming that cinnamon has some sort of upstream xsession file?
Then we could use `services.xserver.displayManager.extraSessionFilePackages` and not create a session manually.

mkg20001 replied:
Yes, they just reference a wrong path
```
Exec=cinnamon-session-cinnamon2d
TryExec=/usr/bin/cinnamon2d
```

jonringer commented:
do you mind cleaning up the commit history? there should rougly be a commit per package change, so multiple fixes should be squashed

`git rebase -i` will be your very dear friend

worldofpeace commented:
> do you mind cleaning up the commit history? there should rougly be a commit per package change, so multiple fixes should be squashed
> 
> `git rebase -i` will be your very dear friend

I haven't suggested a particular commit history because it's very wip.
I'm fine with at least the commits coming in a way where it's changes per package so I can review them properly. In the end, `cinnamon: init at some_version` is fine with me very new package sets.
Though I'm not opposed to it being a commit per package :smile: But at that point I'd hard reset and recreate the `callPackage` file from scratch.

jonringer commented:
actually, I think moving some of these things into other PRs may help on being able to review the changes in a timely manner, some of things that can be refactored to other PRs because they don't rely on other cinnamon packages:

- add to maintainers list
- xapp
- cinnamon-old/cjs
- cinnamon/cjs
- cinnamon/muffin
- cinnamon-control-center
- cinnamon-menus
- cinnamon-screen-saver'

if i were to do a thorough review of this PR, it would be very difficult to adhere to CONTRIBUTING standards

worldofpeace commented:
> actually, I think moving some of these things into other PRs may help on being able to review the changes in a timely manner, some of things that can be refactored to other PRs because they don't rely on other cinnamon packages:
> 
>     * add to maintainers list
> 
>     * xapp
> 
>     * cinnamon-old/cjs
> 
>     * cinnamon/cjs
> 
>     * cinnamon/muffin
> 
>     * cinnamon-control-center
> 
>     * cinnamon-menus
> 
>     * cinnamon-screen-saver'
> 
> 
> if i were to do a thorough review of this PR, it would be very difficult to adhere to CONTRIBUTING standards

It seems we established on https://github.com/NixOS/nixpkgs/pull/74242#issuecomment-558638693.
Though what you've suggested is much better though, it allows reviewers to review less in a sense.
(we've done that for deepin https://github.com/NixOS/nixpkgs/issues/59023, so things do actually get committed)

mkg20001 commented:
So we could keep this PR as the "incubator" and then move things to other PRs for finalization, eventually moving everything from this PR into seperate PRs?

mkg20001 commented:
> While I'm here, why's there a cinnamon-old directory?

It's some older port of cinnamon

I'll look through those to find fixes to issues I have, since I've based off another cinnamon port. (nixos-cinnamon, and cinnamon-old is the old cinnamon from nixpkgs)

mkg20001 commented:
I've created https://github.com/NixOS/nixpkgs/pull/75138 with just the xapps package. Please review :smile:

worldofpeace commented:
@mkg20001, that sounds cool. As long as review doesn't get duplicated, I'll expect you moving your expressions into nixpkgs piece by piece from here.

mkg20001 commented:
I'm still working on the other cinnamon packages in parallel in https://github.com/mkg20001/nixpkgs (this is where I keep all the stuff that is going to be up streamed eventually)

I've stumbled upon https://github.com/NixOS/nixpkgs/issues/9415 because cinnamon fails to load opengl and that seems to be the reason

Also some module fails to find network manager libraries, so some NM stuff is still broken and it can't find adwaita (which package is it in? I've already added gtk3, unsure if I added it to the right package as build input)

Aside from that I've got the desktop working so far without bugs, just the cinnamon spices & nemo extensions are missing

![screenshot](https://user-images.githubusercontent.com/7735145/70788851-1ef49e80-1d92-11ea-84d3-2006002dc223.png)

mkg20001 commented:
merged with https://github.com/NixOS/nixpkgs/pull/97256
